torch/csrc/autograd/engine.cpp

```diff
@@ -651,15 +665,25 @@ auto Engine::execute(const edge_list& roots,
       return graph_task.outstanding_tasks_.load() == 0;
     });
   } else {
-    // Get back to work while we wait for our new graph_task to
-    // complete!
-    // See Note [Reentrant backwards]
-    // If no extra threads remaining, create a new thread for reentrant call
     graph_task.owner_ = worker_device;
-    add_thread_pool_task(&graph_task);
-    graph_task.not_done_.wait(lock, [&graph_task]{
-      return graph_task.outstanding_tasks_.load() == 0;
-    });
+    graph_task.reentrant_depth_ = num_reentrant_threads_[worker_device+1]*max_recursion_depth_+current_depth;
```

Comment:
Oh interesting. I guess the reason you have written the code this way, is that when you make a new worker thread, you "lose" the setting of `current_depth_`. So to get around this problem, you maintain `num_reentrant_threads_` to remember the value for the engine.

I think there is a simpler way to solve this problem though! Since every graph task has a reentrant depth, the computation here should just be `graph_task.reentrant_depth_ = reentrant_depth_ + 1`. Am I mistaken about this? In particular, it doesn't feel like the tracking in `num_reentrant_threads_` should be necessary. I haven't thought about this in as much detail as you, so please let me know if there is a good reason here.

Reply:
We are not keeping track of the graph task while executing the FunctionTask function. So I don't think it'll be available in Engine::execute.